11_FFTBloom/app_resources/fft_convolve_ifft.hlsl

```diff
@@ -0,0 +1,204 @@
+				const vector<scalar_t, 2> zeroKernelVector = kernelChannels[channel].SampleLevel(samplerState[channel], uv, 0);
+				const complex_t<scalar_t> zeroKernel = { zeroKernelVector.x, zeroKernelVector.y };
+				zero = zero * zeroKernel;
+
+				// Do the same for the nyquist coord
+				uv.y += 0.5;
+				const vector<scalar_t, 2> nyquistKernelVector = kernelChannels[channel].SampleLevel(samplerState[channel], uv, 0);
+				const complex_t<scalar_t> nyquistKernel = { nyquistKernelVector.x, nyquistKernelVector.y };
+				nyquist = nyquist * nyquistKernel;
+
+				// Since their IFFT is going to be real, we can pack them back as Z + iN, do a single IFFT and recover them afterwards
+				preloaded[localElementIndex] = zero + rotateLeft<scalar_t>(nyquist);
+
+				// We have set Z + iN for an even element (lower half of the DFT). We must now set conj(Z) + i * conj(N) for an odd element (upper half of DFT)
+				// The logic here is basically the same as in getDFTMirror: we figure out which of our odd elements corresponds to the other thread's
+				// current even element (current even element is `localElementIndex` and our local odd element that's the mirror of the other thread's even element is
+				// `elementToTradeLocalIdx`. Then we get conj(Z) + i * conj(N) from that thread and send our own via a shuffle
+				const complex_t<scalar_t> mirrored = conj(zero) + rotateLeft<scalar_t>(conj(nyquist));
+				vector<scalar_t, 2> mirroredVector = { mirrored.real(), mirrored.imag() };
+				const uint32_t otherElementIdx = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getNablaMirrorIndex(globalElementIndex);
+				const uint32_t otherThreadID = otherElementIdx & (_NBL_HLSL_WORKGROUP_SIZE_ - 1);
+				const uint32_t otherThreadGlobalElementIndex = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | otherThreadID;
+				const uint32_t elementToTradeGlobalIdx = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getNablaMirrorIndex(otherThreadGlobalElementIndex);
+				const uint32_t elementToTradeLocalIdx = elementToTradeGlobalIdx / _NBL_HLSL_WORKGROUP_SIZE_;
+				workgroup::Shuffle<SharedmemAdaptor, vector<scalar_t, 2> >::__call(mirroredVector, otherThreadID, sharedmemAdaptor);
+				preloaded[elementToTradeLocalIdx].real(mirroredVector.x);
+				preloaded[elementToTradeLocalIdx].imag(mirroredVector.y);
+			}
+		}
+		else
+		{
+			for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+			{
+				const uint32_t globalElementIndex = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+				const uint32_t indexDFT = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getDFTIndex(globalElementIndex);
+				const uint32_t bits = pushConstants.numWorkgroupsLog2;
+				const uint32_t y = glsl::bitfieldReverse<uint32_t>(glsl::gl_WorkGroupID().x) >> (32 - bits);
```

Comment:
imho you should make a `bitReverse<typename T, uint16_t Bits NBL_FUNC_REQUIRES(is_scalar_v<T> && Bits<=sizeof(T)*8)` in some namespace that does this for you

P.S. the numWorkgroupsLog2 could be a constexpr because FFT size is hardcoded for both axes already

Reply:
adding to `nbl::hlsl::fft` for now, would it be okay to create a `bit` namespace in `hlsl/bit.hlsl` or is that file reserved to strictly mimic the `std` library?

Reply:
yes that header is reserved for that, but `bit` houses other fun functions like `rotl`
https://en.cppreference.com/w/cpp/numeric/rotl

so due to similarity your function can live there.


bTw you probably want to check `is_scalar` and `is_integer`